hw4/04-templates.js

```diff
@@ -5,60 +5,96 @@ const app = express();
+  if (a.population > b.population) {
+    return -1;
+  }
+  if (a.population < b.population) {
+    return 1;
+  }
+  return 0;
+};
+
+app.get('/', async (req, res) => {
   // render pug template for the index.html file
-
   res.render('index', {
     heading: 'Countries of the World',
     main: 'Welcome to this application. Using the REST Countries API, we will be showing the countries and capitals of the world, the most populous countries in the world, and the number of countries in each region of the world',
   });
 });
-
-app.get('/capitals', (req, res) => {
+app.get('/capitals', async (req, res) => {
   // map the output array to create an array with country names and capitals
   // check for empty data in the output array
-
-  let countries = ['Afghanistan', 'Aland Islands', 'Albania'];
-
+  const countries = (await queryCountries(url)).data;
+  for (let i = 0; i < countries.length; i += 1) {
+    const country = countries[i];
+    const countryName = country.name.common;
+    let capital = '';
+    if (country.capital) [capital] = country.capital;
+    countries[i] = `${countryName} - ${capital}`;
+  }
+  countries.sort(localizedComparator);
   res.render('page', {
     heading: 'Countries and Capitals',
     results: countries,
   });
 });
-
-app.get('/populous', (req, res) => {
+app.get('/populous', async (req, res) => {
   // filter the output array for the countries with population of 50 million or more
   // sort the resulting array to show the results in order of population
   // map the resulting array into a new array with the country name and formatted population
-
-  let populous = ['China', 'India', 'United States of America'];
-
+  const moreThan = 50000000;
+  const countries = (await queryCountries(url)).data;
+  countries.sort(populationComparator);
+  const populous = [];
+  for (let i = 0; i < countries.length; i += 1) {
+    const country = countries[i];
+    const countryName = country.name.common;
+    const { population } = country;
+    if (country.population >= moreThan) {
+      const populationStr = population.toLocaleString('en-US');
+      populous.push(`${countryName} - ${populationStr}`);
+    }
+  }
   res.render('page', {
     heading: 'Most Populous Countries',
     results: populous,
   });
 });
-
-app.get('/regions', (req, res) => {
-  // reduce the output array in a resulting object that will feature the numbers of countries in each region
+app.get('/regions', async (req, res) => {
+  // reduce the output array in a resulting object that will feature the
+  // numbers of countries in each region
   // disregard empty data from the output array
-
-  let regions = ['Asia - 50', 'Europe - 53', 'Africa - 60'];
-
+  const countries = (await queryCountries(url)).data;
+  const regions = [];
+  const regionCountries = {};
+  for (let i = 0; i < countries.length; i += 1) {
+    const country = countries[i];
+    const { region } = country;
+    if (Object.keys(regionCountries).indexOf(region) === -1) {
```

Comment:
This is an interesting way to check if an object has a key.

You can also use the 'in' operator.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in

```javascript
    if (region in regionCountries) {
      regionCountries[region] += 1;
    } else {
      regionCountries[region] = 1;
    }
```

This doesn't show up in Object's methods in MDN docs because it is an operator  so it is hard to know you can do this if that's your main reference.